### PR TITLE
[102X][Backport] DeepAK8 tagger integration

### PR DESCRIPTION
**Pull request by hqucms**
Backport of https://github.com/cms-sw/cmssw/pull/23768. 
Needs external from https://github.com/cms-sw/cmsdist/pull/4326.
The tagger is disabled in the MiniAOD sequence by default to preserve the event content.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test with cms-sw/cmsdist#4326

**Comment by slava77** [comment]
> Reco comparison results: 38 differences found in the comparisons

two discriminants changed quite significantly
and it doesn't look like simple reordering.
e.g. in 136.8311
all_oldvsnew_runjetht2017freminiaod136p8311c_min2 max-2 patjets_slimmedjetsak8__pat_obj___pairdiscrivector__8__second
all_oldvsnew_runjetht2017freminiaod136p8311c_min2 max-2 patjets_slimmedjetsak8__pat_obj___pairdiscrivector__9__second

I guess we have the same problem in the master branch as well. 
Since my tools are simply comparing the discriminants at the same spot, I wasn't easily able to disentangle there the reshuffle made. Note that in the master the deepDoubleB tags were at indices 8 and 9 before and at 32 and 42 after #23768.

This needs to be understood, because this may be a bug affecting the deepDoubleB (at best it's a bug fix).

**Comment by hqucms** [comment]
@slava77 
This is due to the change in https://github.com/cms-sw/cmssw/pull/24505/commits/d7a1c269f02aa60185adf4de15abbf337ee7c563, which is explained in more details in https://github.com/hqucms/cmssw/commit/c04145d1bd9b1671db4d39925fe9a6421d770434#commitcomment-29634182. Before that the deepDoubleB was actually running with the cached `reco::PFCandidate` daughters instead of `pat::PackedCandidate`, even if it was applied on `slimmedJetsAK8`. The difference is generally small for high-pt (>170 GeV) AK8 jets as there it is mainly coming from the different precision between reco and packed objects, but for low-pt (<170) since the daughters are dropped in `slimmedJetsAK8` the difference is quite significant. AFAIK deepDoubleB was developed also for high-pt AK8 jets so the behaviour on those low-pt jets are probably not so relevant anyhow.

**Comment by slava77** [comment]
@jmduarte 
please confirm that the change described in https://github.com/cms-sw/cmssw/pull/24505#issuecomment-420688357 is acceptable, at least for the master branch perspective.


@hqucms 
this backport should strictly reproduce the physics outputs in the default setup. So, some updates are needed in this PR to preserve the old behavior.
In case this feature is important for a possible future re-miniAOD in 102X or at analysis level, it would be nice to leave it configurable.

**Comment by jmduarte** [comment]
@hqucms Are you saying this issue is only causing a difference when running the tagger on RECO inputs? 

How should I test that this? e.g. Should I compare the tagger output running on both RECO and MINIAOD inputs before and after this PR is applied?

Thanks,
Javier

**Comment by hqucms** [comment]
@slava77 
I found a workaround to avoid the change in https://github.com/cms-sw/cmssw/commit/d7a1c269f02aa60185adf4de15abbf337ee7c563. This should allow us to preserve the event content in the backport.

**Comment by hqucms** [comment]
@jmduarte 

If I understand correctly, the way that DeepDoubleB was set up in the MiniAOD sequence (in https://github.com/cms-sw/cmssw/blob/CMSSW_10_2_X/PhysicsTools/PatAlgos/python/slimming/applyDeepBtagging_cff.py#L49-L68), i.e., by updating `slimmedJetsAK8`, was because you wanted to access `pat::PackedCandidate`, which was fully consistent with the training input from MiniAOD files, right? However, due to changes introduced in https://github.com/cms-sw/cmssw/pull/22914, the daughters of `slimmedJetsAK8` are still `reco::PFCandidate` before `slimmedJetsAK8` is written to MiniAOD file (because the daughters are cached in memory), and therefore DeepDoubleB was actually accessing `reco::PFCandidate` when running in the MiniAOD step. 

This PR, specifically https://github.com/cms-sw/cmssw/commit/d7a1c269f02aa60185adf4de15abbf337ee7c563, fixes the change in https://github.com/cms-sw/cmssw/pull/22914 such that the daughters of `slimmedJetsAK8` are always `pat::PackedCandidate`. Then, DeepDoubleB running in MiniAOD step will use `pat::PackedCandidate` which I think is more consistent with the training, right? However, this is then causing some difference for the DeepDoubleB scores produced in the MiniAOD step:

 - for AK8 jets with pt>170, the difference seems very small (I think you also checked in the past using RECO inputs and MINIAOD inputs gave very consistent results, so probably nothing to worry here).

 - for AK8 jets with pt<170, the difference is much larger. This is because the jet daughters are dropped for `slimmedJetsAK8` (https://github.com/cms-sw/cmssw/blob/CMSSW_10_2_X/PhysicsTools/PatAlgos/python/slimming/applySubstructure_cff.py#L306), while previously since we were getting the daughters from the cache, we still have access to all the  `reco::PFCandidate` daughters. 

However, I think only the high pt AK8 jets are of interest for applying DeepDoubleB, right? And since the training was derived with high pt jets, it is most likely not optimal/applicable for these low pt jets anyhow. 

What do you think, @jmduarte?

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/24505	ed815a3
- the backport of https://github.com/cms-sw/cmssw/pull/23768	1acac95 is correct: it is not verbatim to satisfy the no-change policy. The deepAK8 taggers can be enabled in analysis setup.
    - the deepAK8 taggers are disabled (can be enabled by uncommenting some code in PhysicsTools/PatAlgos/python/slimming/applyDeepBtagging_cff.py
    - the changes in DataFormats/Candidate/interface/CompositePtrCandidate.h and DataFormats/PatCandidates/interface/Jet.h are not in this backport PR to preserve the default behavior of the deepDoubleB tagger currently running in production. From the discussion so far in https://github.com/cms-sw/cmssw/pull/24505#issuecomment-420838748 related to these, the change in the master branch seems appropriate.
- jenkins tests pass and comparisons with the baseline show no differences as expected

@fabiocos this PR needs to be merged with https://github.com/cms-sw/cmsdist/pull/4326
IIUC, the dependence on the external becomes essential at runtime only in multi-threaded mode. So, I do not expect any build/run issues in the IBs if the updates are not in sync.

**Comment by hqucms** [comment]
@slava77 

Thank you very much for your review!

> @fabiocos this PR needs to be merged with cms-sw/cmsdist#4326
> IIUC, the dependence on the external becomes essential at runtime only in multi-threaded mode. So, I do not expect any build/run issues in the IBs if the updates are not in sync.

Actually cms-sw/cmsdist#4326 needs to be merged before this PR, because cms-sw/cmsdist#4326 also adds some header files to the MXNet external which are necessary for this PR to compile.

**Comment by slava77** [comment]
On 9/14/18 9:50 AM, Huilin Qu wrote:
> @slava77
> 
> Thank you very much for your review!
> 
>     @fabiocos this PR needs to be merged with cms-sw/cmsdist#4326
>     IIUC, the dependence on the external becomes essential at runtime only in multi-threaded mode. So, I do not expect any build/run issues in the IBs if the updates are not in sync.
> 
> Actually cms-sw/cmsdist#4326 needs to be merged before this PR, because cms-sw/cmsdist#4326 also adds some header files to the MXNet external which are necessary for this PR to compile.
> 

good to know.
Thanks for the correction, it may have averted a minor disaster.

File: PhysicsTools/PatAlgos/python/slimming/applyDeepBtagging_cff.py [hunk 1]
```diff
@@ -58,19 +58,20 @@ def applyDeepBtagging( process, postfix="" ) :
...
        elSource = cms.InputTag('slimmedElectrons'),
        rParam = 0.8,
-       jetCorrections = ('AK8PFPuppi', cms.vstring(['L1FastJet', 'L2Relative', 'L3Absolute']), 'None'),
+       jetCorrections = ('AK8PFPuppi', cms.vstring(['L2Relative', 'L3Absolute']), 'None'),
        btagDiscriminators = [
           'pfDeepDoubleBJetTags:probQ',
           'pfDeepDoubleBJetTags:probH',
        ],
-       postfix = 'SlimmedAK8DeepDoubleB'+postfix,
+#        ] + pfDeepBoostedJetTagsAll,  # uncomment it to test DeepBoostedJet
```
Comment (fabiocos) [inline PhysicsTools/PatAlgos/python/slimming/applyDeepBtagging_cff.py]: @slava77 I understand that this is the line that would trigger the production of the new collection, for the erst the old one is renamed (DeepDoubleB -> DeppTags) but it is not a persistent collection

Comment (slava77) [reply PhysicsTools/PatAlgos/python/slimming/applyDeepBtagging_cff.py]: I'm not sure I understand ```for the erst ```.

The change of the name in the transient collection names starts at L49 and the postfix above also updates other transient collection names.

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge

**Pull request merged**